// @mui/x-charts/ScatterChart/ScatterPlot.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ScatterPlot = ScatterPlot;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _clsx = _interopRequireDefault(require("clsx"));
var _styles = require("@mui/material/styles");
var _Scatter = require("./Scatter");
var _useScatterSeries = require("../hooks/useScatterSeries");
var _hooks = require("../hooks");
var _useZAxis = require("../hooks/useZAxis");
var _seriesConfig = require("./seriesConfig");
var _BatchScatter = require("./BatchScatter");
var _scatterClasses = require("./scatterClasses");
var _jsxRuntime = require("react/jsx-runtime");
const ScatterPlotRoot = (0, _styles.styled)('g', {
  name: 'MuiScatterPlot',
  slot: 'Root'
})();

/**
 * Demos:
 *
 * - [Scatter](https://mui.com/x/react-charts/scatter/)
 * - [Scatter demonstration](https://mui.com/x/react-charts/scatter-demo/)
 *
 * API:
 *
 * - [ScatterPlot API](https://mui.com/x/api/charts/scatter-plot/)
 */
function ScatterPlot(props) {
  const {
    slots,
    slotProps,
    onItemClick,
    renderer,
    className,
    classes: inClasses
  } = props;
  const seriesData = (0, _useScatterSeries.useScatterSeriesContext)();
  const {
    xAxis,
    xAxisIds
  } = (0, _hooks.useXAxes)();
  const {
    yAxis,
    yAxisIds
  } = (0, _hooks.useYAxes)();
  const {
    zAxis,
    zAxisIds
  } = (0, _useZAxis.useZAxes)();
  const classes = (0, _scatterClasses.useUtilityClasses)({
    classes: inClasses
  });
  if (seriesData === undefined) {
    return null;
  }
  const {
    series,
    seriesOrder
  } = seriesData;
  const defaultXAxisId = xAxisIds[0];
  const defaultYAxisId = yAxisIds[0];
  const defaultZAxisId = zAxisIds[0];
  const DefaultScatterItems = renderer === 'svg-batch' ? _BatchScatter.BatchScatter : _Scatter.Scatter;
  const ScatterItems = slots?.scatter ?? DefaultScatterItems;
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(ScatterPlotRoot, {
    className: (0, _clsx.default)(classes.root, className),
    children: seriesOrder.map(seriesId => {
      const {
        id,
        xAxisId,
        yAxisId,
        zAxisId,
        color,
        hidden
      } = series[seriesId];
      if (hidden) {
        return null;
      }
      const colorGetter = _seriesConfig.scatterSeriesConfig.colorProcessor(series[seriesId], xAxis[xAxisId ?? defaultXAxisId], yAxis[yAxisId ?? defaultYAxisId], zAxis[zAxisId ?? defaultZAxisId]);
      const xScale = xAxis[xAxisId ?? defaultXAxisId].scale;
      const yScale = yAxis[yAxisId ?? defaultYAxisId].scale;
      return /*#__PURE__*/(0, _jsxRuntime.jsx)(ScatterItems, (0, _extends2.default)({
        xScale: xScale,
        yScale: yScale,
        color: color,
        colorGetter: colorGetter,
        series: series[seriesId],
        onItemClick: onItemClick,
        slots: slots,
        slotProps: slotProps
      }, slotProps?.scatter), id);
    })
  });
}
process.env.NODE_ENV !== "production" ? ScatterPlot.propTypes = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  /**
   * Callback fired when clicking on a scatter item.
   * @param {MouseEvent} event Mouse event recorded on the `<svg/>` element.
   * @param {ScatterItemIdentifier} scatterItemIdentifier The scatter item identifier.
   */
  onItemClick: _propTypes.default.func,
  /**
   * The type of renderer to use for the scatter plot.
   * - `svg-single`: Renders every scatter item in a `<circle />` element.
   * - `svg-batch`: Batch renders scatter items in `<path />` elements for better performance with large datasets, at the cost of some limitations.
   *                Read more: https://mui.com/x/react-charts/scatter/#performance
   *
   * @default 'svg-single'
   */
  renderer: _propTypes.default.oneOf(['svg-batch', 'svg-single']),
  /**
   * The props used for each component slot.
   * @default {}
   */
  slotProps: _propTypes.default.object,
  /**
   * Overridable component slots.
   * @default {}
   */
  slots: _propTypes.default.object
} : void 0;
