// @mui/x-charts/LineChart/lineClasses.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getLineUtilityClass = getLineUtilityClass;
exports.useUtilityClasses = exports.lineClasses = void 0;
var _generateUtilityClass = _interopRequireDefault(require("@mui/utils/generateUtilityClass"));
var _composeClasses = _interopRequireDefault(require("@mui/utils/composeClasses"));
var _generateUtilityClasses = _interopRequireDefault(require("@mui/utils/generateUtilityClasses"));
function getLineUtilityClass(slot) {
  return (0, _generateUtilityClass.default)('MuiLineChart', slot);
}
const lineClasses = exports.lineClasses = (0, _generateUtilityClasses.default)('MuiLineChart', ['area', 'line', 'mark', 'markAnimate', 'highlight', 'areaPlot', 'linePlot', 'markPlot']);
const useUtilityClasses = options => {
  const {
    skipAnimation,
    classes
  } = options ?? {};
  const slots = {
    area: ['area'],
    line: ['line'],
    mark: ['mark', !skipAnimation && 'markAnimate'],
    highlight: ['highlight'],
    areaPlot: ['areaPlot'],
    linePlot: ['linePlot'],
    markPlot: ['markPlot']
  };
  return (0, _composeClasses.default)(slots, getLineUtilityClass, classes);
};
exports.useUtilityClasses = useUtilityClasses;
