// @mui/x-charts/PieChart/seriesConfig/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.pieSeriesConfig = void 0;
var _seriesProcessor = _interopRequireDefault(require("./seriesProcessor"));
var _getColor = _interopRequireDefault(require("./getColor"));
var _legend = _interopRequireDefault(require("./legend"));
var _tooltip = _interopRequireDefault(require("./tooltip"));
var _seriesLayout = _interopRequireDefault(require("./seriesLayout"));
var _getSeriesWithDefaultValues = _interopRequireDefault(require("./getSeriesWithDefaultValues"));
var _tooltipPosition = _interopRequireDefault(require("./tooltipPosition"));
var _keyboardFocusHandler = _interopRequireDefault(require("./keyboardFocusHandler"));
var _identifierSerializer = require("../../internals/identifierSerializer");
const pieSeriesConfig = exports.pieSeriesConfig = {
  colorProcessor: _getColor.default,
  seriesProcessor: _seriesProcessor.default,
  seriesLayout: _seriesLayout.default,
  legendGetter: _legend.default,
  tooltipGetter: _tooltip.default,
  tooltipItemPositionGetter: _tooltipPosition.default,
  getSeriesWithDefaultValues: _getSeriesWithDefaultValues.default,
  keyboardFocusHandler: _keyboardFocusHandler.default,
  identifierSerializer: _identifierSerializer.identifierSerializerSeriesIdDataIndex
};
