// @mui/x-charts/PieChart/seriesConfig/seriesLayout.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _getPercentageValue = require("../../internals/getPercentageValue");
var _getPieCoordinates = require("../getPieCoordinates");
const seriesLayout = (series, drawingArea) => {
  const seriesLayoutRecord = {};
  for (const seriesId of series.seriesOrder) {
    const {
      innerRadius,
      outerRadius,
      arcLabelRadius,
      cx: cxParam,
      cy: cyParam
    } = series.series[seriesId];
    const {
      cx,
      cy,
      availableRadius
    } = (0, _getPieCoordinates.getPieCoordinates)({
      cx: cxParam,
      cy: cyParam
    }, {
      width: drawingArea.width,
      height: drawingArea.height
    });
    const outer = (0, _getPercentageValue.getPercentageValue)(outerRadius ?? availableRadius, availableRadius);
    const inner = (0, _getPercentageValue.getPercentageValue)(innerRadius ?? 0, availableRadius);
    const label = arcLabelRadius === undefined ? (inner + outer) / 2 : (0, _getPercentageValue.getPercentageValue)(arcLabelRadius, availableRadius);
    seriesLayoutRecord[seriesId] = {
      radius: {
        available: availableRadius,
        inner,
        outer,
        label
      },
      center: {
        x: drawingArea.left + cx,
        y: drawingArea.top + cy
      }
    };
  }
  return seriesLayoutRecord;
};
var _default = exports.default = seriesLayout;
