// @mui/x-charts/RadarChart/FocusedRadarMark.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FocusedRadarMark = FocusedRadarMark;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _styles = require("@mui/material/styles");
var _useFocusedItem = require("../hooks/useFocusedItem");
var _useRadarSeriesData = require("./RadarSeriesPlot/useRadarSeriesData");
var _jsxRuntime = require("react/jsx-runtime");
function FocusedRadarMark(props) {
  const theme = (0, _styles.useTheme)();
  const focusedItem = (0, _useFocusedItem.useFocusedItem)();
  const seriesCoordinates = (0, _useRadarSeriesData.useRadarSeriesData)(focusedItem?.seriesId);
  if (!focusedItem || focusedItem.type !== 'radar' || seriesCoordinates.length === 0) {
    return null;
  }
  const point = seriesCoordinates[0].points[focusedItem.dataIndex];
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("rect", (0, _extends2.default)({
    fill: "none",
    stroke: (theme.vars ?? theme).palette.text.primary,
    strokeWidth: 2,
    x: point.x - 6,
    y: point.y - 6,
    width: 2 * 6,
    height: 2 * 6,
    rx: 3,
    ry: 3
  }, props));
}
