// @mui/x-charts/esm/Gauge/GaugeValueArc.js
'use client';

import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
const _excluded = ["className"],
  _excluded2 = ["cx", "cy", "startAngle", "endAngle", "cornerRadius", "innerRadius", "outerRadius", "skipAnimation"];
import * as React from 'react';
import PropTypes from 'prop-types';
import { styled } from '@mui/material/styles';
import clsx from 'clsx';
import { useSkipAnimation } from "../hooks/useSkipAnimation.js";
import { useAnimateGaugeValueArc } from "../hooks/animation/useAnimateGaugeValueArc.js";
import { useGaugeState } from "./GaugeProvider.js";
import { gaugeClasses } from "./gaugeClasses.js";
import { jsx as _jsx } from "react/jsx-runtime";
const StyledPath = styled('path', {
  name: 'MuiGauge',
  slot: 'ValueArc'
})(({
  theme
}) => ({
  fill: (theme.vars || theme).palette.primary.main
}));
function GaugeValueArc(_ref) {
  let {
      className
    } = _ref,
    other = _objectWithoutPropertiesLoose(_ref, _excluded);
  const {
    value,
    valueMin,
    valueMax,
    startAngle,
    endAngle,
    outerRadius,
    innerRadius,
    cornerRadius,
    cx,
    cy
  } = useGaugeState();
  if (value === null) {
    return null;
  }
  const valueAngle = startAngle + (value - valueMin) / (valueMax - valueMin) * (endAngle - startAngle);
  return /*#__PURE__*/_jsx(AnimatedGaugeValueArc, _extends({}, other, {
    className: clsx(gaugeClasses.valueArc, className),
    cx: cx,
    cy: cy,
    startAngle: startAngle,
    endAngle: valueAngle,
    cornerRadius: cornerRadius,
    innerRadius: innerRadius,
    outerRadius: outerRadius
  }));
}
process.env.NODE_ENV !== "production" ? GaugeValueArc.propTypes = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  skipAnimation: PropTypes.bool
} : void 0;
export { GaugeValueArc };
function AnimatedGaugeValueArc(_ref2) {
  let {
      cx,
      cy,
      startAngle,
      endAngle,
      cornerRadius,
      innerRadius,
      outerRadius,
      skipAnimation: inSkipAnimation
    } = _ref2,
    other = _objectWithoutPropertiesLoose(_ref2, _excluded2);
  const skipAnimation = useSkipAnimation(inSkipAnimation);
  const animatedProps = useAnimateGaugeValueArc({
    startAngle,
    endAngle,
    cornerRadius,
    innerRadius,
    outerRadius,
    skipAnimation
  });
  return /*#__PURE__*/_jsx(StyledPath, _extends({}, animatedProps, {
    transform: `translate(${cx}, ${cy})`
  }, other));
}
process.env.NODE_ENV !== "production" ? AnimatedGaugeValueArc.propTypes = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  cornerRadius: PropTypes.number.isRequired,
  cx: PropTypes.number.isRequired,
  cy: PropTypes.number.isRequired,
  endAngle: PropTypes.number.isRequired,
  innerRadius: PropTypes.number.isRequired,
  outerRadius: PropTypes.number.isRequired,
  skipAnimation: PropTypes.bool,
  startAngle: PropTypes.number.isRequired
} : void 0;
