// @mui/x-charts/context/ChartsProvider/ChartsProvider.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChartsProvider = ChartsProvider;
exports.defaultSeriesConfig = void 0;
var React = _interopRequireWildcard(require("react"));
var _useCharts = require("../../internals/store/useCharts");
var _ChartsContext = require("./ChartsContext");
var _useChartCartesianAxis = require("../../internals/plugins/featurePlugins/useChartCartesianAxis");
var _useChartTooltip = require("../../internals/plugins/featurePlugins/useChartTooltip");
var _useChartInteraction = require("../../internals/plugins/featurePlugins/useChartInteraction");
var _useChartZAxis = require("../../internals/plugins/featurePlugins/useChartZAxis");
var _useChartHighlight = require("../../internals/plugins/featurePlugins/useChartHighlight/useChartHighlight");
var _seriesConfig = require("../../BarChart/seriesConfig");
var _seriesConfig2 = require("../../LineChart/seriesConfig");
var _seriesConfig3 = require("../../PieChart/seriesConfig");
var _seriesConfig4 = require("../../ScatterChart/seriesConfig");
var _jsxRuntime = require("react/jsx-runtime");
const defaultSeriesConfig = exports.defaultSeriesConfig = {
  bar: _seriesConfig.barSeriesConfig,
  scatter: _seriesConfig4.scatterSeriesConfig,
  line: _seriesConfig2.lineSeriesConfig,
  pie: _seriesConfig3.pieSeriesConfig
};

// For consistency with the v7, the cartesian axes are set by default.
// To remove them, you can provide a `plugins` props.
const defaultPlugins = [_useChartZAxis.useChartZAxis, _useChartTooltip.useChartTooltip, _useChartInteraction.useChartInteraction, _useChartCartesianAxis.useChartCartesianAxis, _useChartHighlight.useChartHighlight];
function ChartsProvider(props) {
  const {
    children,
    plugins = defaultPlugins,
    pluginParams = {},
    seriesConfig = defaultSeriesConfig
  } = props;
  const {
    contextValue
  } = (0, _useCharts.useCharts)(plugins, pluginParams, seriesConfig);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_ChartsContext.ChartsContext.Provider, {
    value: contextValue,
    children: children
  });
}
